In [1]:
# this mounts your Google Drive to the Colab VM.
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

# enter the foldername in your Drive where you have saved the unzipped
# assignment folder, e.g. 'ece697ls/assignments/assignment3/'
FOLDERNAME = "ECE662-Lab1"
#assert FOLDERNAME is not None, "[!] Enter the foldername."

# now that we've mounted your Drive, this ensures that
# the Python interpreter of the Colab VM can load
# python files from within it.
import os, sys
ROOT = "/content/drive/My Drive/ECE662-Lab1"
os.chdir(ROOT)
if ROOT not in sys.path:
    sys.path.append(ROOT)

print("Current dir:", os.getcwd())
print("ECE662 folder contents:", os.listdir("ece662"))

#%cd /content/drive/My\ Drive/ECE662-Lab1

Mounted at /content/drive
Current dir: /content/drive/My Drive/ECE662-Lab1
ECE662 folder contents: ['data_utils.py', 'vis_utils.py', 'setup.py', '__init__.py', 'gradient_check.py', 'fast_layers.py', 'im2col.py', 'pruning_helper.py', '__pycache__', 'notebook_images', 'models', 'datasets', 'classifiers', 'solver.py', 'optim.py', 'im2col_cython.pyx', 'build', 'im2col_cython.c', 'im2col_cython.cpython-312-x86_64-linux-gnu.so', 'layer_utils.py', 'layers.py']


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
!ls "/content/drive/My Drive/ECE662-Lab1/"

In [ ]:
!git config --global user.name "Lris00994"
!git config --global user.email "ruili@umass.edu"


import os, getpass
os.environ['GITHUB_PAT'] = getpass.getpass('Enter your GitHub token: ')

In [ ]:
%cd /content
!rm -rf Lab1-Rui-Li
!git clone https://Lris00994:$GITHUB_PAT@github.com/Lris00994/Lab1-Rui-Li.git
%cd Lab1-Rui-Li

In [ ]:
%%bash
cat > .gitignore << 'EOF'
.ipynb_checkpoints/
*.zip
*.tmp
*.log
__pycache__/
.DS_Store
EOF

git add .gitignore
git commit -m "chore: add .gitignore" || true

In [ ]:
%cd /content/Lab1-Rui-Li


!cp "/content/drive/MyDrive/ECE662-Lab1/Q6-Pruning.ipynb" .


!git add Q6-Pruning.ipynb
!git commit -m "Q6: Rruning+Unstructured Pruning+Structured Pruning"
!git push origin main

# Pruning

After learning, neural networks have modified and learned a set of parameters to perform our classification task. However, such parameters are costly to maintain and do not hold the same importance.

Wouldn't it be great could optimize our resource usage by dropping less important values ? This is where pruning comes into play.

Pruning is a technique that cuts off parameters/structures from a model to increase sparcity and decrease overall model size, similar to cutting leafs or branches from bushes and trees. This process can lead to smaller memory consumption with minimal accuracy reduction. Moreover, pruning the network may also provide a speedup since there will be less operations being performed.

The pruning process can be performed during the end of an epoch of training or after training is complete. Experimenting to find out which way works the best is part of the fun !

In [9]:
from ece662.pruning_helper import test_model, load_model
from ece662.data_utils import get_CINIC10_data
import os

Below we will load a pre-trained model for you to work on. If you prefer, you can save your own model from the previous Tensorflow/Pytorch task and load it here.

In [11]:
#This code may take a while to execute as it is training a network form scratch

data = get_CINIC10_data()
mode = 'torch'#torch or tensorflow

test_data = [data['X_test'],data['y_test']]

path = os.path.join('/content/drive/My Drive/{}'.format(FOLDERNAME), f"ece662/models/{mode}.model")


model = load_model(path,mode=mode)
test_model(model,test_data,mode=mode)

Test Acc: 0.5085


## Unstructured Pruning

Unstructured Pruning is usually related to the pruning of weights in neural networks. The general idea is to select a set of weights according to a policy and setting them up to zero.

Common policies are random weight selection or selecting the smallers weights.
Unstructured Pruning can be performed in one or multiple layers within the same network.

Altough in theory Unstructured Pruning should decrease the number of operations performed during execution there should be explicit support within the framework or hardware to bypass such operations, otherwise it will just operated over zero.

### Perform Pruning

Using the model trained in the previous step using pytorch, perform unstructured pruning in the weights of the model by removing x% of the smallest weights.

*   Increment global pruning by 10% until reaching total of 80% pruned weights
*   Perform inference at the end of each pruning and observe the impact into the accuracy.


Note: The percentages are related to the entire model, not per layer.



In [14]:
from importlib import reload
import ece662.pruning_helper
reload(ece662.pruning_helper)

from ece662.pruning_helper import test_model


In [15]:
################################################################################
# TODO: Perform unstructured Pruning over the trained model using 3 different
# prunning percentages.
################################################################################
# *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
import torch.nn.utils.prune as prune
import torch.nn as nn

pruning_percentages = [10, 20, 30, 40, 50, 60, 70, 80]
accuracies = []

for pct in pruning_percentages:
    parameters_to_prune = []
    for name, module in model.named_modules():
        if isinstance(module, nn.Conv2d) or isinstance(module, nn.Linear):
            parameters_to_prune.append((module, 'weight'))
    prune.global_unstructured(
        parameters_to_prune,
        pruning_method=prune.L1Unstructured,
        amount=pct / 100.0
    )

    acc = test_model(model, test_data, mode=mode)
    print(f"Pruned {pct}% - Accuracy: {acc:.4f}")
    accuracies.append((pct, acc))


# *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
################################################################################
#                              END OF YOUR CODE                                #
################################################################################

Test Acc: 0.5011
Pruned 10% - Accuracy: 0.5011
Test Acc: 0.4988
Pruned 20% - Accuracy: 0.4988
Test Acc: 0.4871
Pruned 30% - Accuracy: 0.4871
Test Acc: 0.4454
Pruned 40% - Accuracy: 0.4454
Test Acc: 0.3767
Pruned 50% - Accuracy: 0.3767
Test Acc: 0.2383
Pruned 60% - Accuracy: 0.2383
Test Acc: 0.1701
Pruned 70% - Accuracy: 0.1701
Test Acc: 0.1667
Pruned 80% - Accuracy: 0.1667


## Inline Question 1:

What happened with the accuracy as the % of pruning increased ?
Why was that the case?


## Answer:

[As the percentage of pruning increased, the accuracy gradually dropped. This happened because we were removing more and more weights from the network, including ones that were important for making good predictions. With too many weights gone, the model couldn’t learn or remember enough, so performance got worse.]

## Structured Pruning

Structured Pruning consists of removing a bigger chunk of the network parameters at the same time. Instead of removing only a few weights, it is commonplace to remove entire neurons.

For example, in Convolutional Layers, removing filters can be beneficial to improve performance as it greatly decreases the amount of computation performed. However, some of these changes may affect output dimensions which may be carried over to other parts of the network. Therefore, when performing structured pruning one must always be aware of which parameters are going to be affected.

Using the previously trained model in the CINIC-10, perform Structured Prunning only in the Convolution layers of the DNN.

In [18]:
################################################################################
# TODO: Perform unstrucuted Pruning over the trained model using 3 different
# prunning percentages.
################################################################################
# *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
import torch
import torch.nn.utils.prune as prune
import copy

original_model = model

prune_percentages = [0.3, 0.5, 0.7]
accuracies = []

for pct in prune_percentages:
    model_copy = copy.deepcopy(original_model)


    for name, module in model_copy.named_modules():
        if isinstance(module, torch.nn.Conv2d):
            prune.ln_structured(module, name="weight", amount=pct, n=2, dim=0)

    acc = test_model(model_copy, test_data, mode=mode)
    print(f"Structured Pruned {int(pct*100)}% - Accuracy: {acc:.4f}")
    accuracies.append((pct, acc))


# *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
################################################################################
#                              END OF YOUR CODE                                #
################################################################################

Test Acc: 0.1667
Structured Pruned 30% - Accuracy: 0.1667
Test Acc: 0.1667
Structured Pruned 50% - Accuracy: 0.1667
Test Acc: 0.1667
Structured Pruned 70% - Accuracy: 0.1667


## Inline Question 2:

What is the difference between performing Structured Pruning vs Dropout ?
Why would it be beneficial to perform both techniques when developing a Neural Network?


## Answer:

[Structured pruning means removing whole parts of the model (like filters), so it makes the model smaller and faster. But in my case, pruning too much made accuracy drop a lot-around 16.67%.

Dropout just randomly turns off neurons during training to avoid overfitting, but it doesn’t change the model’s size.

Using both：dropout makes training better, pruning makes the model smaller after training.]
